examples/trials/mnist-tfv2/mnist.py

```diff
@@ -0,0 +1,164 @@
+from tensorflow.keras.layers import (Conv2D, Dense, Dropout, Flatten, MaxPool2D)
+from tensorflow.keras.optimizers import Adam
+
+import nni
+
+_logger = logging.getLogger('mnist_example')
+
+# Uncomment following lines to print metrics to stdout:
+# logging.basicConfig()
+# _logger.setLevel(logging.INFO)
+
+
+class MnistModel(Model):
+    """
+    LeNet-5 Model with customizable hyper-parameters
+    """
+    def __init__(self, conv_size, hidden_size, dropout_rate):
+        """
+        Initialize hyper-parameters.
+
+        Parameters
+        ----------
+        conv_size : int
+            Kernel size of convolutional layers.
+        hidden_size : int
+            Dimensionality of last hidden layer.
+        dropout_rate : float
+            Dropout rate between two fully connected (dense) layers, to prevent co-adaptation.
+        """
+        super().__init__()
+        self.conv1 = Conv2D(32, kernel_size=conv_size, activation='relu')
+        self.pool1 = MaxPool2D(pool_size=2)
+        self.conv2 = Conv2D(filters=64, kernel_size=conv_size, activation='relu')
+        self.pool2 = MaxPool2D(pool_size=2)
+        self.flatten = Flatten()
+        self.fc1 = Dense(units=hidden_size, activation='relu')
+        self.dropout = Dropout(rate=dropout_rate)
+        self.fc2 = Dense(units=10, activation='softmax')
+
+    def call(self, x):
+        """Override ``Model.call`` to build LeNet-5 model."""
+        x = self.conv1(x)
+        x = self.pool1(x)
+        x = self.conv2(x)
+        x = self.pool2(x)
+        x = self.flatten(x)
+        x = self.fc1(x)
+        x = self.dropout(x)
+        return self.fc2(x)
+
+
+class ReportIntermediates(Callback):
+    """
+    Callback class for reporting intermediate accuracy metrics.
+
+    This callback sends accuracy to NNI framework every 100 steps,
+    so you can view the learning curve on web UI.
+
+    If an assessor is configured in experiment's YAML file,
+    it will use these metrics for early stopping.
+
+    Please note that most assessors expect all trials to have same number of intermediate results
+    (if its not early-stopped).
+    """
+    def __init__(self, batch_size):
+        """
+        Initializer. Batch_size is used to output fixed number of intermediate results.
+        """
+        self._batch_size = batch_size
+        self._epoch = 0
+        self._steps = 0
+
+    def on_batch_end(self, batch, logs=None):
+        """Reports accuracy for every 10000 steps"""
+        self._steps += self._batch_size
+        while self._steps >= 10000:
+            self._steps -= 10000
+            _logger.info(
+                'Epoch: %s  Batch: %s  Intermediate accuracy: %s',
+                self._epoch, batch, logs['accuracy']
+            )
+            nni.report_intermediate_result(logs['accuracy'])  # send accuracy to NNI framework
+
+    def on_epoch_end(self, epoch, logs=None):
+        """To log current epoch"""
+        self._epoch += 1
+        self._steps = 0
```

Comment:
Normally we do not use step and epoch at the same time, for small dataset like MNIST, it is OK to just use epoch for simplicity.

Reply:
Because the batch size is generated by tuner, an epoch may have different number of mini-batches.
So if we report intermediate result every mini-batch, trials will have different length of intermediate results.
I'm trying to find a more elegant way...

Reply:
It's more natural to report validation accuracy periodically instead of training accuracy, which requires doing validation periodically.

Reply:
you can report intermediate results on epoch end.

Reply:
Or should we use batch size to calculate epoch number instead?
It looks unfair to me if trials have different epochs, but this is what TF v1 example did (if I understood it correctly).

Reply:
If the whole story is about steps (there is no such thing as epochs), then it's fair. If you are counting epochs, then perhaps you should forget about steps.

Reply:
I just found out that we can simply send intermediate result on epoch end...
Anyway we still need to decide whether to fix epoch number or to fix mini-batch number (`steps_per_epoch`).

Reply:
we do not need to fix epoch number,  it's fine to be unfair,  or we can simply remove epochs from hyper parameters.

Reply:
Sorry I missed some conversation because the page was not auto-refreshed...
Updating.